### ISSUE TITLE
[feature request] keybindings which take keys as arguments

### ISSUE THREAD

**Issue by herrsimon** [tridactyl/tridactyl#1985]
Hello,

would it be possible to implement dynamic keybindings, i.e. keybindings which take one (or multiple) keys as arguments which can be used in the code that follows? What I mean would for example be a `bindwithargs` command of the form 

`bindarg [key] [numargs] [code]`

so that for example

`bindarg a 1 js alert(ARG1)`

would bind `a` to output whatever key is pressed afterwards (so pressing `ab` would output `b` etc.). Pentadactyl had this functionality and it came in very handy using custom scheme handlers. In my case, I used this with orgmodes org-protocol so that I could capture the current url using different templates.

**Comment by glacambre** [comment]
I believe `bind a gobble 1 js -p alert(JS_ARG)` does what you want. It's pretty hard to find in the docs. If you searched for it, what keywords did you use? We should add them to `:gobble`'s documentation to make it easier to find.

**Comment by herrsimon** [comment]
Perfect, this works like a charm! I was searching for "dynamic keybinding", "pass arguments to keybinding" and "dynamically pass arguments to javascript".

I think you should add your answer not only to `:gobble` but also to `:bind`.
Also, it would be helpful to add to the documentation of `js` that if one uses `js -p`, the argument can not only be piped but alternatively added as the last argument in case it contains no whitespace. However, this is quite confusing, because 

`composite echo some multiword argument | js -p alert(JS_ARG);`

works as expected (alerts 'some multiword argument'), but

`js -p alert(JS_ARG) "some multiword argument"`

fails. So maybe one should add details on how the last argument to `js -p` is determined (I guess it takes everything from the last whitespace of the command line until the end).

In any case, thanks a lot for your help, this was the last functionality I needed to finally migrate from palemoon/pentadactyl to firefox/tridactyl.

**Comment by DivineMK** [comment]
I wonder if this documentation can be updated or this issue can be linked to in tridactyl documentation?

**Comment by bovine3dom** [comment]
yeah, the issue is open because we need to improve the documentation. In most cases the `-d[eof character]` version is more intuitive, e.g.

```
js -d¹ window.alert(JS_ARGS.join(" "))¹ hello this is a test
composite echo hello this is a test | js -d¹ window.alert(JS_ARGS.join(" "))¹
```

both work as expected

**Comment by GHolk** [comment]
The answer in 2019 does not work today ( `bind a gobble 1 js -p alert(JS_ARG)` )
because gobble pass the key as the first arguments,
so `js -p alert(JS_ARG)` become `js <KEYS> -p alert(JS_ARG)` .

Maybe we should add an option to make gobble pass keys as last argument?

**Comment by bovine3dom** [comment]
that sounds reasonable. i don't remember why we made that change (it's the opposite to what composite does...)

**Comment by GHolk** [comment]
I checked the git log.

In beginning, gobble mode does not support args.

0a0c31230e511bdab0781a8efe5e9fa671e095c6 add gobble mode and quickmarks

Then someone adds args support, which pass them after the position of char.

ded80d2fa9946aeca18218f90242f91d21c06174 Add bindkey command

To pass arguments at last, wrap the whole command string as a intermediate command.

```
command js-p-alert js -p alert(JS_ARG)
bind a gobble 1 js-p-alert
```
